### PR TITLE
feat(tls): add TlsStream support

### PR DESCRIPTION
**Pull request by compressed**
Here's my attempt to add TLS support to tarpc. I had first started with two tls-specific client and server files, but that seemed like a bit too much duplication so I merged those changes into the existing client and server files. The changes are really quite small, despite the large diff. The main changes are for the server it needs to take in a `TlsAcceptor` and for the client it needs to take in the domain you want to connect to along with a `TlsConnector`; both of these tls types are provided by the `native-tls` crate, which I have re-exported.

I needed to do some `replace`ing in the `Cargo.toml` because the released version of `native-tls` doesn't have the macos tls-related functions exposed that I needed for testing. This will also give a superfluous warning when compiling without the tls feature (`package replacement is not used: ...`). Once there is a new native-tls release, we can get rid of that replace.

For the `service!` macro I was able to eliminate almost all of the duplication via some helper macros.

I was also able to re-use all the tests with just some minimal refactoring.

The examples currently won't build right now when the `tls` feature is specified. I was thinking we could use some conditional compilation there like I did in the tests as well. I'll work on that next.

Let me know what you think!

-----
Commit message below:

When `-- features tls` is specified for tarpc, RPC communication will
occur over a `TlsStream<TcpStream>` instead of a `TcpStream`.

The `service!` macro has been refactored to limit the amount of code
duplication between the non-tls and tls versions.

In addition, the client and server logic has also been refactored to
limit the amount of duplication between the two code paths.

The functional tests have been refactored to use a common set of
functions for constructing the client and server structs so that all
the tests are shared across non-tls and tls test runs.

### CODE REVIEW

**Comment by tikue** [comment]
This is awesome, thanks so much! We'll start reviewing soon.

File: Cargo.toml [hunk 1]
```diff
@@ -25,19 +27,32 @@ take = "0.1"
...
+[target.'cfg(target_os = "macos")'.dev-dependencies]
+security-framework = "0.1"
 
 [features]
+default = []
+tls = ["tokio-tls", "native-tls"]
 unstable = ["serde/unstable"]
 
+[replace]
+"tokio-core:0.1.2" = { git = "https://github.com/tokio-rs/tokio-core" }
```
Comment (tikue) [inline Cargo.toml:54]: This was updated to 0.1.3.

File: README.md [hunk 2]
```diff
@@ -131,6 +131,115 @@ fn main() {
...
+     let acceptor = TlsAcceptor::builder(pkcs12).unwrap().build().unwrap();
+     let client_cx = TlsClientContext::try_new("foobar.com").unwrap();
+
+     (acceptor, client_cx)
+}
+
+fn main() {
+    let addr = "localhost:10000";
+    let (acceptor, client_cx) = tls_context();
+    let _server = HelloServer.listen(addr, acceptor);
```
Comment (tikue) [inline README.md:181]: I always intended there to be an additional config param for all the miscellaneous things one might want to tweak. This seems like a good time to introduce that. Also, in the config struct, could the tls acceptor be optional `Option<TlsAcceptor>`? I think we should allow the mixing of TLS and non-TLS services, which I believe isn't possible as this PR currently stands?

Comment (tikue) [reply README.md:181]: And I think the same should be applied to the client side, so maybe in the respective modules there could be `client::Config` and `server::Config`. Those modules could be made public.

File: README.md [hunk 3]
```diff
@@ -131,6 +131,115 @@ fn main() {
...
+     let client_cx = TlsClientContext::try_new("foobar.com").unwrap();
+
+     (acceptor, client_cx)
+}
+
+fn main() {
+    let addr = "localhost:10000".first_socket_addr();
+    let mut core = reactor::Core::new().unwrap();
+    let (acceptor, client_cx) = tls_context();
+    HelloServer.listen_with(addr, core.handle(), acceptor).unwrap();
```
Comment (tikue) [inline README.md:233]: If we go with the config struct approach, the handle should become a field of that as well. Not sure what'd be better, to have one client config struct, or one for each of sync, futures.

File: src/errors.rs [hunk 4]
```diff
@@ -107,3 +107,9 @@ pub enum WireError<E> {
 pub trait SerializableError: StdError + Deserialize + Serialize + Send + 'static {}
 
 impl<E: StdError + Deserialize + Serialize + Send + 'static> SerializableError for E {}
+
+#[cfg(feature = "tls")]
+/// Convert `native_tls::Error` to `std::io::Error`
+pub fn native2io(e: ::native_tls::Error) -> io::Error {
```
Comment (tikue) [inline src/errors.rs:113]: nit: prefer `native_to_io`

**Review by tikue — CHANGES_REQUESTED** [review]
From what I've read so far, this looks great! Some initial comments.

File: src/server.rs [hunk 5]
```diff
@@ -20,57 +20,112 @@ use tokio_service::NewService;
...
-                        Error = io::Error> + Send + 'static,
-          Req: Deserialize + 'static,
-          Resp: Serialize + 'static,
-          E: Serialize + 'static
-{
-    let (tx, rx) = futures::oneshot();
-    REMOTE.spawn(move |handle| Ok(tx.complete(listen_with(addr, new_service, handle.clone()))));
-    ListenFuture { inner: rx }
-}
+cfg_if! {
```
Comment (tikue) [inline src/server.rs:23]: if the `tls_acceptor` arg is replaced with a config struct, then we can probably constrain the `cfg_if!` to a much smaller surface of the code. It's fine if the config struct is empty (for now) when tls isn't enabled.

File: README.md (hunk 2 above)
Comment (compressed) [reply README.md:181]: Yep, you're right. It doesn't allow mixing TLS and non-TLS currently. That's what I was asking about on gitter before. However, I do think it's good to try to allow this.

I really like the idea of the `Config` structs for client and server. That change should then unify the API then across both stream types and I believe then we'd only need one version of the macro.

Let me give this a shot and see where I end up. Thinking about the client, I believe we'll need to add another generic parameter to indicate the stream type...

**Comment by compressed** [comment]
Closing this one in place of a new one on `google/tarpc`.

**Pull request closed without merging**